### ISSUE TITLE
Synchronize portfolio after buying in live mode

### ISSUE THREAD

**Issue by Knifespirit** [enigmampc/catalyst#111]
Dear Catalyst Maintainers,

Before I tell you about my issue, let me describe my environment:

# Environment

* Operating System: ubuntu 16.04
* Python Version: 2.7
* Python Bitness: 64
* How did you install Catalyst: pip
* Python packages: 
`alembic==0.9.6
bcolz==0.12.1
boto3==1.5.1
botocore==1.8.15
Bottleneck==1.2.1
ccxt==1.10.400
certifi==2017.11.5
chardet==3.0.4
click==6.7
contextlib2==0.5.5
cyordereddict==1.0.0
Cython==0.27.3
decorator==4.1.2
docutils==0.14
empyrical==0.2.1
enigma-catalyst==0.4.0
futures==3.2.0
idna==2.6
int-date==0.1.8
intervaltree==2.1.0
jmespath==0.9.3
Logbook==1.1.0
lru-dict==1.1.6
Mako==1.0.7
MarkupSafe==1.0
multipledispatch==0.4.9
networkx==2.0
numexpr==2.6.4
numpy==1.13.3
pandas==0.19.2
pandas-datareader==0.5.0
patsy==0.4.1
pkg-resources==0.0.0
python-dateutil==2.6.1
python-editor==1.0.3
pytz==2017.3
requests==2.18.4
requests-file==1.4.2
requests-ftp==0.3.1
s3transfer==0.1.12
scipy==1.0.0
six==1.11.0
sortedcontainers==1.5.9
SQLAlchemy==1.1.15
statsmodels==0.8.0
stockstats==0.2.0
tables==3.4.2
toolz==0.9.0
urllib3==1.22
`

Now that you know a little about me, let me tell you about the issue I am
having:

# Description of Issue
I am running a live bot on binance and evertime the bot buys something it will crash

* What did you expect to happen?
The bot would continue after buying

* What happened instead?

`    for capital_change_packet in every_bar(dt):
  File "/home/catalyst/catalyst-venv/local/lib/python2.7/site-packages/catalyst/gens/tradesimulation.py", line 137, in every_bar
    handle_data(algo, current_data, dt_to_use)
  File "/home/catalyst/catalyst-venv/local/lib/python2.7/site-packages/catalyst/utils/events.py", line 216, in handle_data
    dt,
  File "/home/catalyst/catalyst-venv/local/lib/python2.7/site-packages/catalyst/utils/events.py", line 235, in handle_data
    self.callback(context, data)
  File "/home/catalyst/catalyst-venv/local/lib/python2.7/site-packages/catalyst/exchange/exchange_algorithm.py", line 641, in handle_data
    cash, positions_value = self.synchronize_portfolio()
  File "/home/catalyst/catalyst-venv/local/lib/python2.7/site-packages/catalyst/exchange/exchange_algorithm.py", line 523, in synchronize_portfolio
    raise ValueError('Cash on exchanges is lower than the algo.')
ValueError: Cash on exchanges is lower than the algo.
`

Here is how you can reproduce this issue on your machine:

## Reproduction Steps

1. Run the mean_reversion_simple on live mode
2. The bot will find something to buy
3. It will crash

## What steps have you taken to resolve this already?
Couldnt think of anything to solvce this

Sincerely,
Vincent

**Comment by Knifespirit** [comment]
I just tried commenting the ValueError in my catalyst, This seems to do the trick. It doesnt crash anymore.

When i look at the output of my console I notice it will take binance takes a few minutes to give the new cash value. 

Maybe there should be some room of error, where the bot waits till the cash is the same as the in the exchange and only errors when it takes to long

**Comment by camanio** [comment]
I have the same error when trying to buy two or more assets in my portfolio algo. By the way, how I can save and load actual positions when I restart the bot?

**Comment by Knifespirit** [comment]
When running live mode it already creates a file that keeps track of this. The file is named after the namespace of the algo. Myself i had problems with this because when i adjusted the algo and traded again it would get incorrect cash and postition numbers. (I also made a issue about this right after this issue)

**Comment by fredfortier** [comment]
It seems like you are saying that the algo raises a ValueError when it tries to fetch the cash balance from Binance?

Catalysts does fetch a balance after every frame. That's how it can keep stats like p&l up to date. It's always aware of the latest price of a position. It also ensures that the cash and positions in the algo are equal or larger than in the exchange account.

When there is an error fetching data from an exchange, Catalyst should retry at least five times before giving up. Do you have an error trace of this error so I can have a better idea of where it is failing? We do have an algo trading live on Binance. We have not experienced this issue so far.

**Comment by Knifespirit** [comment]
Right now i cannot test it. I will look at it later to see if i can add some more details. 

From what i remember the algo crashed right after buying without first rechecking.

**Comment by camanio** [comment]
Some comments about this bug. I met it too in Bittrex live mode. Just after the deal cash didn't updated (I checked exchange balance and  exposure_stats.csv in live_algos folder). Deal executed perfectly, I saw it in orders history at exchange.

Moreover, there are additional bugs: 
1. If you buy multiple assets in same time, current price updated only for first asset. 
2. If you try to get current prices for list of assets (for example: prices = data.current(context.assets, 'price') where context.assets is a list), you can see differnt behaivor in live and backtest modes. In backtest mode all is fine (each column in DataFrame contains prices for corresponding asset), but in live mode all columns filled with prices of the first asset.

Think, it will help to locate the bug.

**Comment by fredfortier** [comment]
Yes, that helps. We’ll resolve this asap.

**Comment by quocble** [comment]
I'm also experiencing this on Binance. I executing multiple sells and multiple buys on different assets in one go... then the sync fails.

**Comment by fredfortier** [comment]
The original issue (cash synchronization) is resolved (in the develop branch for now, but we'll do a release tonight). Looking into @camanio 's comment.

**Comment by fredfortier** [comment]
@camanio what do you mean by this: "If you buy multiple assets in same time, current price updated only for first asset.". Are you talking about the stats being printed at the bottom of the algo output for each frame. If you, how do you record the two prices?

**Comment by fredfortier** [comment]
@camanio we've improved the stats printing to support multiple assets. See commit attached to this issue for example. The command `data.current(assets, 'price)` does return the correct values for the specified assets in live and backtest. This is from the `develop` branch but we will release tonight.

Here is the stats output for this example:
```
                                                                                                                                    symbol  amount  cost_basis  last_sale_price         price
period_close              starting_cash ending_cash portfolio_value pnl         long_exposure short_exposure orders transactions                                                             
2018-01-04 01:25:00+00:00 0.01          -0.11475568 0.00968858      -0.00031142 0.12444426    0              1      1              eth_btc       2  0.06237784       0.06222213    0.06222213
                                                                                                                    1             eth_usdt       0  0.00000000       0.00000000  936.12540593

```

**Comment by fredfortier** [comment]
Please reopen if it's still happening in version 0.4.2.

**Comment by camanio** [comment]
@fredfortier Thanks a lot! All stats calcs perfectly, I just tested it in live mode at Bittrex.

**Comment by camanio** [comment]
@fredfortier Hello! This problem arise again. Please, try to reproduce these steps in live mode:
1. Open some positions, two positions in different assets are enough. You will see correct stats update and at this stage all OK.
2. Interrupt algo. It is usual situation - exception may arise due some problems with exchange, you wish fix some bugs in algo, etc. 
3. Run the algo again (you must decrease capital_base and turn off condition in exchange.exchange.py, line 693) and try to sell any asset (for example, you can buy & sell assets by timer). You will see very strange behaivor:
- sell order accepted by exchange
- total assets remain same (catalyst does not see the deal)

I trace execution in debugger - sync with exchange is OK, but portfolio does not update. Without interruption catalyst works well. May be something wrong with save/restore procedure?

By the way, can you allow restarting algos without hacking of catalyst?

**Comment by fredfortier** [comment]
Thanks @camanio ! I got it. We'll issue a new release when fixed.

If by restarting algos, you mean clearing their state, use this command:
```
catalyst clean-algo -n [your algo namespace]
```

**Comment by fredfortier** [comment]
Fixed. Basically, Zipline uses `PerformanceTracker` object to keep track of positions. This object is an attribute of three other objects, two of which are being pickled to keep the algo state. When restoring the state, I had to propagate one of the saved `PerformanceTracker` to all three objects.

I'll take care of one more things before releasing this.

**Comment by camanio** [comment]
Thanks, @fredfortier !  I'll test the patch ASAP.

About restarting an algo. If I not mistaken, clean-algo delete all information about algo state. But what's about restarting the same algo with its current state? If I run clean-algo, I must close all postions manually via exchange web interface, for example and after that run the algo again. But I want to restart the algo automatically - for example, in case of some troubles at exchange side (as a rule, catalyst crashes if exchange does not respond for some time). Now I'm using for these cases try - exception block around run_algorithm function and small hack mentioned above (disable check for minimum capital amount). But I think it is very dirty way and this wornderfull platform can support restarting out-of-the-box.

**Comment by camanio** [comment]
@fredfortier ! I just tested your patch in live mode at Bittrex. There is a part of logs:

[2018-01-08 11:26:15.009718] INFO: port_eaa_btc_v5p: BAR 2018-01-08 11:25:00+00:00 BUY: TradingPair(833710 [salt_btc]) @0.00080461 amount: 32.38477931483457 (0)
[2018-01-08 11:26:15.010303] INFO: CCXT: adjusted order amount 32.3847793148 to 32.38477931 based on lot size
[2018-01-08 11:26:15.272044] INFO: exchange_algorithm: statistics for the last 1 minutes:
                           starting_cash  ending_cash  portfolio_value  pnl  long_exposure  short_exposure  orders  transactions  leverage
period_close                                                                                                                              
2018-01-08 11:26:00+00:00           0.03         0.03             0.03  0.0              0               0       0             0       0.0
[2018-01-08 11:26:15.518703] INFO: CCXT: order deb9112f-b01e-4593-8339-e1ac4a3e5e44 executed in full: 32.38477931 salt_btc
[2018-01-08 11:26:15.731238] WARNING: Exchange: detected lower balance for salt_btc on bittrex: 0.0 < 32.38477931, updating position amount
[2018-01-08 11:26:15.734103] INFO: exchange_algorithm: portfolio balances, cash: 0.0281014, positions: 0.0260571172806
[2018-01-08 11:26:15.815477] INFO: exchange_algorithm: statistics for the last 1 minutes:
                           starting_cash  ending_cash  portfolio_value         pnl  long_exposure  short_exposure  orders  transactions    leverage
period_close                                                                                                                                       
2018-01-08 11:27:00+00:00           0.03   0.00381983       0.00381983 -0.02618017              0               0       2             1  0.87267233
[2018-01-08 11:27:01.199748] INFO: exchange_algorithm: portfolio balances, cash: 0.02863124, positions: 0.0
[2018-01-08 11:27:01.277624] INFO: exchange_algorithm: statistics for the last 1 minutes:
                           starting_cash  ending_cash  portfolio_value         pnl  long_exposure  short_exposure  orders  transactions  leverage
period_close                                                                                                                                     
2018-01-08 11:28:00+00:00           0.03   0.00381983       0.00381983 -0.02618017              0               0       0             0       0.0
[2018-01-08 11:28:00.644121] INFO: exchange_algorithm: portfolio balances, cash: 0.02863124, positions: 0.0
[2018-01-08 11:28:00.698451] INFO: exchange_algorithm: statistics for the last 1 minutes:
                           starting_cash  ending_cash  portfolio_value         pnl  long_exposure  short_exposure  orders  transactions  leverage
period_close                                                                                                                                     
2018-01-08 11:29:00+00:00           0.03   0.00381983       0.00381983 -0.02618017              0               0       0             0       0.0

As you can see, algo succefully bought some salt_btc (I checked the deal via exchange web interface), but apropriate amount does not appear at stats table. The same problem was before the patch after restarting an algo, but now it arises at first run.

**Comment by Ericxgao** [comment]
Not sure if this is related - but for some reason my portfolio amounts are also way off.

    [2018-01-10 09:08:01.418589] INFO: goose1: 2018-01-10 09:08:00+00:00: buying - price: 7.7e-06
    [2018-01-10 09:08:01.418728] INFO: goose1: 7.93e-06: limit - stop: 7.585e-06
    [2018-01-10 09:08:01.548843] INFO: CCXT: adjusted order amount 389.61038961 to 389.0 based on lot size
    [2018-01-10 09:08:01.706259] INFO: exchange_algorithm: statistics for the last 1 minutes:
                               starting_cash  ending_cash  portfolio_value  pnl  long_exposure  short_exposure  orders  transactions      price  price_change   cash
    period_close                                                                                                            
    2018-01-10 09:09:00+00:00          0.003        0.003            0.003  0.0              0               0       0             0  0.0000077    0.04904632  0.003
    [2018-01-10 09:08:01.710049] WARNING: exchange_algorithm: unable save stats locally: 'price'
    [2018-01-10 09:09:00.872266] INFO: CCXT: order 16542720 executed in full: 389.0 trx_btc
    [2018-01-10 09:09:01.154222] WARNING: Exchange: detected lower balance for trx_btc on binance: 388.679 < 389.0, updating position amount
    [2018-01-10 09:09:01.155661] INFO: exchange_algorithm: got totals from exchanges, cash: 0.06450654 positions: 0.00292528
    [2018-01-10 09:09:02.345206] INFO: goose1: 2018-01-10 09:09:00+00:00: selling - price: 7.52e-06 - bought at: 7.7e-06
    [2018-01-10 09:09:02.478747] INFO: CCXT: adjusted order amount 388.679 to 388.0 based on lot size
    [2018-01-10 09:09:02.645483] INFO: exchange_algorithm: statistics for the last 1 minutes:
    period_close              starting_cash ending_cash portfolio_value pnl        long_exposure short_exposure orders transactions price      price_change cash
    2018-01-10 09:10:00+00:00 0.003         0.003       0.00592287      0.00292287 0.00292287    0              1      1            0.00000752 0.02452316   0.003  trx_btc  388.679         0.0       0.00000752

I sold at a loss, but my portfolio value is now double?

**Comment by fredfortier** [comment]
We haven't pushed the patch yet as we had to focus on something else. We'll
do this today. Which exchange are you trading with?

**Comment by Ericxgao** [comment]
@fredfortier 

Binance. If I look closer at the output, it seems like the problem is that the ending cash is not deducted after the trade, thus giving the illusion that I have double the portfolio value.

It thinks I didn't spend any cash when buying the coin.

**Comment by fredfortier** [comment]
Okay. I trade with Binance as well so this must have triggered a specific condition. I believe that you are commenting in the right issue as it could be a side-effect of the same thing. Did this happen with a new algo (either first run or after running `catalyst clean-algo -n [your algo]`) or after resuming an existing run? We've just released version 0.4.4 which fixes this issue. If you can, please give it another shot and see if it resolved your error condition. I'll keep this open until I hear back.

**Comment by Ericxgao** [comment]
I tried both deleting the files from the /live_algos/ directory and running that - same issue both times.

Let me try it again with 0.4.4.

**Comment by Ericxgao** [comment]
@fredfortier 

No luck unfortunately. Clean algo - here's the output again.

    [2018-01-10 21:15:00.602993] INFO: exchange_algorithm: portfolio balances, cash: 0.06742138, positions: 0.0
    [2018-01-10 21:15:02.100542] INFO: goose1: 2018-01-10 21:15:00+00:00: buying - price: 1.15e-05
    [2018-01-10 21:15:02.101162] INFO: goose1: 1.187e-05: limit - stop: 1.1315e-05
    [2018-01-10 21:15:02.233005] INFO: CCXT: adjusted order amount 260.869565217 to 260.0 based on lot size
    [2018-01-10 21:15:02.383263] INFO: exchange_algorithm: statistics for the last 1 minutes:
                               starting_cash  ending_cash  portfolio_value  pnl  long_exposure  short_exposure  orders  transactions      price  price_change   cash
    period_close                                                                                                                                                    
    2018-01-10 21:16:00+00:00          0.003        0.003            0.003  0.0              0               0       0             0  0.0000115   -0.00862069  0.003
    [2018-01-10 21:16:00.594331] INFO: CCXT: order 11727630 executed in full: 260.0 xvg_btc
    [2018-01-10 21:16:00.943841] WARNING: Exchange: detected lower balance for xvg_btc on binance: 259.74 < 260.0, updating position amount
    [2018-01-10 21:16:00.946030] INFO: exchange_algorithm: portfolio balances, cash: 0.06443138, positions: 0.0029926
    [2018-01-10 21:16:02.241796] INFO: exchange_algorithm: statistics for the last 1 minutes:
    period_close              starting_cash ending_cash portfolio_value pnl        long_exposure short_exposure orders transactions price      price_change cash                                               
    2018-01-10 21:17:00+00:00 0.003         0.003       0.00598961      0.00298961 0.00298961    0              1      1            0.00001151 -0.00775862  0.003  xvg_btc  259.74         0.0       0.00001151

**Comment by Ericxgao** [comment]
After the sell (at a loss) my portfolio value is up.

    [2018-01-10 21:37:00.911173] INFO: exchange_algorithm: portfolio balances, cash: 0.06146468, positions: 0.002940425631
    [2018-01-10 21:37:02.190715] INFO: exchange_algorithm: statistics for the last 1 minutes:
    symbol    amount  cost_basis  last_sale_price
    period_close              starting_cash ending_cash portfolio_value pnl        long_exposure short_exposure orders transactions price    price_change cash                                                 
    2018-01-10 21:38:00+00:00 0.003         0.003       0.00594043      0.00294043 0.00294043    0              0      0            0.089102 -0.01642565  0.003  eth_btc  0.032967         0.0         0.089193
    [2018-01-10 21:38:00.545120] INFO: exchange_algorithm: portfolio balances, cash: 0.06146468, positions: 0.002925590481
    [2018-01-10 21:38:02.382941] INFO: goose3: 2018-01-10 21:38:00+00:00: selling - price: 0.088895 - bought at: 0.0899
    [2018-01-10 21:38:02.520413] INFO: CCXT: adjusted order amount 0.032967 to 0.032 based on lot size
    [2018-01-10 21:38:02.716929] INFO: exchange_algorithm: statistics for the last 1 minutes:
    symbol    amount  cost_basis  last_sale_price
    period_close              starting_cash ending_cash portfolio_value pnl        long_exposure short_exposure orders transactions price    price_change cash                                                 
    2018-01-10 21:39:00+00:00 0.003         0.003       0.00592559      0.00292559 0.00292559    0              0      0            0.088895 -0.01871067  0.003  eth_btc  0.032967         0.0         0.088743
    [2018-01-10 21:39:00.936860] INFO: CCXT: order 50516559 executed in full: 0.032 eth_btc
    [2018-01-10 21:39:01.222285] WARNING: Exchange: detected lower balance for eth_btc on binance: 0.000967 < 0.000967, updating position amount
    [2018-01-10 21:39:01.222915] INFO: exchange_algorithm: portfolio balances, cash: 0.06430072, positions: 8.5675233e-05
    [2018-01-10 21:39:02.558247] INFO: exchange_algorithm: statistics for the last 1 minutes:
    symbol    amount  cost_basis  last_sale_price
    period_close              starting_cash ending_cash portfolio_value pnl        long_exposure short_exposure orders transactions price  price_change cash                                                 
    2018-01-10 21:40:00+00:00 0.003         0.003       0.00308568      0.00008568 0.00008568    0              1      1            0.0885 -0.02307098  0.003  eth_btc  0.000967         0.0         0.088599

**Comment by Ericxgao** [comment]
This only occurs when simulate_orders=False - the values are correct when simulate_orders=True on Binance as well.

**Comment by fredfortier** [comment]
Thanks for the update. I don't get it. I'm running live algos against
Binance and my cash is updating just fine. Can you paste your catalyst or
`run_algorithm` command? Did you set your algo base_currency to btc?

**Comment by Ericxgao** [comment]
My run_algorithm is in the code:

    run_algorithm(
        capital_base=0.003,
        initialize=initialize,
        handle_data=handle_data,
        analyze=analyze,
        exchange_name='binance',
        live=True,
        algo_namespace=NAMESPACE,
        base_currency='btc',
        live_graph=False,
        simulate_orders=True
    )

**Comment by fredfortier** [comment]
Ok thanks. It looks good. I’ll run some tests with the same market and look
for the issue.

**Comment by Ericxgao** [comment]
@fredfortier 

Here's some more info.

I thought maybe my algo was causing it, but even with the simple buy and hold the numbers are off:

This is the algo:
    import os
    import tempfile
    import time

    import numpy as np
    import pandas as pd
    import talib
    from logbook import Logger

    from catalyst import run_algorithm
    from catalyst.api import symbol, record, order_target_percent, get_open_orders
    from catalyst.exchange.utils.stats_utils import extract_transactions
    from catalyst.utils.paths import ensure_directory

    NAMESPACE = 'goose7'
    log = Logger(NAMESPACE)

    from catalyst.api import order, record, symbol

    def initialize(context):
        context.asset = symbol('trx_btc')

    def handle_data(context, data):
        order_target_percent(
                    context.asset, 1
                )
        record(btc = data.current(context.asset, 'price'))

    if __name__ == '__main__':
        run_algorithm(
            capital_base=0.003,
            initialize=initialize,
            handle_data=handle_data,
            exchange_name='binance',
            live=True,
            algo_namespace=NAMESPACE,
            base_currency='btc',
            live_graph=False,
            simulate_orders=False
        )

And my output:

    [2018-01-11 00:48:20.434855] WARNING: run_algo: Catalyst is currently in ALPHA. It is going through rapid development and it is subject to errors. Please use carefully. We encourage you to report any issue on GitHub: https://github.com/enigmampc/catalyst/issues
    [2018-01-11 00:48:23.436079] INFO: run_algo: running algo in live-trading mode
    [2018-01-11 00:48:23.437371] INFO: exchange_bundle: pricing data for [u'btc_usdt'] not found in range 2015-03-01 00:00:00+00:00 to 2018-01-10 00:00:00+00:00, updating the bundles.
    Ingesting daily price data for btc_usdt on poloniex
    [2018-01-11 00:48:23.936927] WARNING: Loader: Refusing to download new treasury data because a download succeeded at 2018-01-11 00:43:21.307477+00:00.
    [2018-01-11 00:48:23.939454] INFO: exchange_algorithm: initialized trading algorithm in live mode
    [2018-01-11 00:48:24.502005] INFO: exchange_algorithm: portfolio balances, cash: 0.06135499, positions: 0.0
    [2018-01-11 00:48:24.632055] INFO: CCXT: adjusted order amount 356.718192628 to 356.0 based on lot size
    [2018-01-11 00:48:24.938390] INFO: exchange_algorithm: statistics for the last 1 minutes:
                               starting_cash  ending_cash  portfolio_value  pnl  long_exposure  short_exposure  orders  transactions         btc
    period_close                                                                                                                                
    2018-01-11 00:49:00+00:00          0.003        0.003            0.003  0.0              0               0       0             0  0.00000841
    [2018-01-11 00:49:01.123454] INFO: CCXT: order 17147337 executed in full: 356.0 trx_btc
    [2018-01-11 00:49:01.402316] INFO: exchange_algorithm: portfolio balances, cash: 0.05836103, positions: 0.0029904
    [2018-01-11 00:49:01.532101] INFO: CCXT: adjusted order amount 1.14285714286 to 1.0 based on lot size
    [2018-01-11 00:49:01.809692] INFO: exchange_algorithm: statistics for the last 1 minutes:
    symbol  amount  cost_basis  last_sale_price
    period_close              starting_cash ending_cash portfolio_value pnl       long_exposure short_exposure orders transactions btc                                                     
    2018-01-11 00:50:00+00:00 0.003         0.003       0.0059904       0.0029904 0.0029904     0              1      1            0.00000841  trx_btc   356.0         0.0        0.0000084
    [2018-01-11 00:49:20.797595] INFO: exchange_algorithm: Interruption signal detected 2, exiting the algorithm
    [2018-01-11 00:49:20.798182] INFO: exchange_algorithm: Exiting the algorithm.


Another note - this was working fine until yesterday so some state got messed up... is there a way to maybe just do a clean reinstall? I upgraded from 0.4.3 to 0.4.4 as well.

**Comment by fredfortier** [comment]
You could give re-install a shot although I'd prefer getting to the bottom
of it. Which version of Python are you using?

**Comment by Ericxgao** [comment]
Using 2.7

**Comment by fredfortier** [comment]
Yes, that's the most tested version. I'll try your specific algo, see if
the problem is reproducible and go from there.

**Comment by fredfortier** [comment]
Okay I found the issue. Basically, it arises when placing market orders. At first glance, it seems like Binance unexpectedly forgets to give us an executed price completing the order. I will document this as a separate issue For now, I'd recommend using limit orders, which is generally a good practice anyway.

I made some adjustments to your algo:
```
from logbook import Logger

from catalyst import run_algorithm
from catalyst.api import order_target_percent

NAMESPACE = 'goose7'
log = Logger(NAMESPACE)

from catalyst.api import record, symbol


def initialize(context):
    context.asset = symbol('trx_btc')


def handle_data(context, data):
    price = data.current(context.asset, 'price')
    record(btc=price)

    # Only ordering if it does not have any position to avoid trying some
    # tiny orders with the leftover btc
    pos_amount = context.portfolio.positions[context.asset].amount
    if pos_amount > 0:
        return

    # Adding a limit price to workaround an issue with performance
    # calculations of market orders
    order_target_percent(
        context.asset, 1, limit_price=price * 1.05
    )


if __name__ == '__main__':
    run_algorithm(
        capital_base=0.003,
        initialize=initialize,
        handle_data=handle_data,
        exchange_name='binance',
        live=True,
        algo_namespace=NAMESPACE,
        base_currency='btc',
        live_graph=False,
        simulate_orders=False,
    )

```

**Comment by fredfortier** [comment]
I also noticed related issue #151. For the time being, your perf calculation will look more pessimistic, assuming that orders are executing at your limit price.

**Comment by Ericxgao** [comment]
Do you know if I need to do limit orders for my sells as well?

**Comment by fredfortier** [comment]
@Ericxgao the issue is fixed. You should now get an execution price for market orders as well. I think that limit orders are a good idea most of the time, unless you're reading from the orderbook targeting specific offers. Fixed on 0.4.6 please re-open if you still notice an issue.
